experimental/framework/audience-heartbeat/src/audienceWithHeartBeat.ts

```diff
@@ -0,0 +1,102 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+import { EventEmitter } from "events";
+import { IAudience } from "@fluidframework/container-definitions";
+import { IClient, MessageType } from "@fluidframework/protocol-definitions";
+import { IInboundSignalMessage } from "@fluidframework/runtime-definitions";
+import { IFluidDataStoreRuntime } from "@fluidframework/datastore-definitions";
+import { IFluidAudienceWithHeartBeat } from "./interfaces";
+
+/**
+ * This class wraps implements heart-beat over Audience and emits event
+ * in case we are not listening from any client for 5 continuous heart-beats
+ * or we started getting heart-beat from a client who is not present in Audience.
+ */
+export class AudienceWithHeartBeat extends EventEmitter implements IFluidAudienceWithHeartBeat {
+    private readonly frequency: number;
+    private readonly audienceHeartBeat: Map<string, Date> = new Map();
+    private readonly audience: IAudience ;
+    private readonly runtime: IFluidDataStoreRuntime ;
+    private timer: any = undefined;
+
+    /**
+     * Creates a AudienceWithHeartBeat object.
+     * @param audience - Audience.
+     * @param frequency - heartbeat frequency in milliseconds.
+     */
+    constructor(
+        audience: IAudience,
+        runtime: IFluidDataStoreRuntime,
+        frequency: number = 30000) {
```

Comment:
>30000 [](http://example.com/codeflow?start=28&length=5)

Do we intent to have every readonly client to send ping every 30 seconds? That will quickly get to  a lot of signal as more clients join the session.  

I am wondering whether we are solving right problem here.  I believe your intent is to make Audience more reliable so that it can be relied on it to pick a read only client to do some work.   But there are alternatives to sove that problem.  Can we solve the problem more directly?

I wonder would it suffice if the client that is supposed to do the work send the singal when it is doing the work, and the other validate it that way? 

The key of validating it is understanding the error rate of signals, which can be measured

Reply:
Currently, I am also not sure about frequency that's why I thought of to keep it in params to keep it configurable, may be after some load test we will be to know what should be default frequency.

To make sure low load I added enableHeartBeat and disableHeartBeat so the moment we have one member in quorum we can disable it and when there is no member in quorum we can enable it again.

Reply:
In general, we should avoid bloating the network traffic that scales up by the number of client.     Does all the client really need to know that all the read client is there for your scenario?   As I indicated, I believe you only need to make sure that one that you chose to do the work is alive for your scenario.   That's O(1) vs. O(n)

Reply:
In this PR, I am try to propose a generic scenario to improve reliability of audience.

For leader only signal I think we should handle it, FFX level as leader selection may vary.

Reply:
We need to avoid the multiple leader problem too as far as we can. For that signalling to all clients will be needed. @curtisman please let us know if you feel something different?